src/control/cmd/daos_server/storage.go

```diff
@@ -35,97 +35,184 @@ type storagePrepareCmd struct {
 	HelperLogFile string `short:"l" long:"helper-log-file" description:"Log debug from daos_admin binary."`
 }
 
-func (cmd *storagePrepareCmd) Execute(args []string) error {
-	prepNvme, prepScm, err := cmd.Validate()
-	if err != nil {
+type errs []error
+
+func (es *errs) add(err error) error {
+	if es == nil {
+		return errors.New("nil errs")
+	}
+	*es = append(*es, err)
+
+	return nil
+}
+
+type nvmePrepFn func(storage.BdevPrepareRequest) (*storage.BdevPrepareResponse, error)
+
+func (cmd *storagePrepareCmd) prepNvme(scanErrors *errs, prep nvmePrepFn) error {
+	if doPrep, _, err := cmd.Validate(); err != nil || !doPrep {
 		return err
 	}
 
-	if cmd.HelperLogFile != "" {
-		if err := os.Setenv(pbin.DaosAdminLogFileEnvVar, cmd.HelperLogFile); err != nil {
-			cmd.log.Errorf("unable to configure privileged helper logging: %s", err)
+	op := "Prepare"
+	if cmd.Reset {
+		op = "Reset"
+	}
+
+	cmd.log.Info(op + " locally-attached NVMe storage...")
+
+	if cmd.TargetUser == "" {
+		runningUser, err := user.Current()
+		if err != nil {
+			return errors.Wrap(err, "couldn't lookup running user")
 		}
+		cmd.TargetUser = runningUser.Username
 	}
 
-	// This is a little ugly, but allows for easier unit testing.
-	// FIXME: With the benefit of hindsight, it seems apparent
-	// that we should have made these Execute() methods thin
-	// wrappers around more easily-testable functions.
-	if cmd.scs == nil {
-		cmd.scs = server.NewStorageControlService(cmd.log, config.DefaultServer().Engines)
+	// Prepare NVMe access through SPDK
+	if _, err := prep(storage.BdevPrepareRequest{
+		HugePageCount: cmd.NrHugepages,
+		TargetUser:    cmd.TargetUser,
+		PCIAllowList:  cmd.PCIAllowList,
+		PCIBlockList:  cmd.PCIBlockList,
+		Reset_:        cmd.Reset,
+		EnableVMD:     true, // vmd will be prepared if available
+	}); err != nil {
+		return scanErrors.add(err)
 	}
 
-	op := "Preparing"
-	if cmd.Reset {
-		op = "Resetting"
+	return nil
+}
+
+type scmPrepFn func(storage.ScmPrepareRequest) (*storage.ScmPrepareResponse, error)
+
+func (cmd *storagePrepareCmd) prepScm(scanErrors *errs, prep scmPrepFn) error {
+	if _, doPrep, err := cmd.Validate(); err != nil || !doPrep {
+		return err
 	}
 
-	scanErrors := make([]error, 0, 2)
+	if cmd.NrNamespacesPerSocket == 0 {
+		return errors.New("(-S|--scm-ns-per-socket) should be set to at least 1")
```

Comment:
should we also impose an upper limit? for example <=4... i do see a case to support up to four (e.g. 4x 100GbE Ethernet links per socket), but >4 is unlikely to be needed anytime soon.